vcd/resource_vcd_edgegateway.go

```diff
@@ -0,0 +1,201 @@
+					"delete_force": &schema.Schema{
+						Type:        schema.TypeBool,
+						Required:    true,
+						ForceNew:    true,
+						Description: "TBD: no description in the API, and undefined behavior in practice",
+					},
+					"delete_recursive": &schema.Schema{
+						Type:        schema.TypeBool,
+						Required:    true,
+						ForceNew:    true,
+						Description: "TBD: no description in the API, and undefined behavior in practice",
+					},
+			*/
+			"advanced_networking": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "True if the gateway uses advanced networking",
+			},
+			"backing_config": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Configuration of the vShield edge VM for this gateway. One of: compact, full",
+			},
+			"ha_enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "Enable high availability on this edge gateway",
+			},
+			"external_networks": &schema.Schema{
+				Type:        schema.TypeList,
+				Required:    true,
+				ForceNew:    true,
+				Description: "A list of external networks to be used by the edge gateway",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"default_gateway": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				ForceNew:    true,
+				Description: "External network to be used as default gateway. Its name must be included in 'external_networks'. An empty value will skip the default gateway",
+			},
+			"distributed_routing": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "If advanced networking enabled, also enable distributed routing",
+			},
+		},
+	}
+}
+
+// Creates a new edge gateway from a resource definition
+func resourceVcdEdgeGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] edge gateway creation initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	rawExternalNetworks := d.Get("external_networks").([]interface{})
+	var externalNetworks []string
+	for _, en := range rawExternalNetworks {
+		externalNetworks = append(externalNetworks, en.(string))
+	}
+
+	var gwCreation = govcd.EdgeGatewayCreation{
+		ExternalNetworks:          externalNetworks,
+		Name:                      d.Get("name").(string),
+		OrgName:                   d.Get("org").(string),
+		VdcName:                   d.Get("vdc").(string),
+		Description:               d.Get("description").(string),
+		BackingConfiguration:      d.Get("backing_config").(string),
+		AdvancedNetworkingEnabled: d.Get("advanced_networking").(bool),
+		DefaultGateway:            d.Get("default_gateway").(string),
```

Comment:
Maybe it is wise to check that default_gateway isn't missing and is one of externalNetworks?

Reply:
This check is done in go-vcloud-director. The creation function returns an error if default gateway is not in the external networks list.
An empty `default_gateway` is valid input, as it means "no default gateway"

Reply:
It's already checked in govcd

Reply:
That's fine too.